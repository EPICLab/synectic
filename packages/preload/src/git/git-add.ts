import type {PathLike} from 'fs-extra';
import {relative} from 'path';
import {execute} from '../io';
import {getWorktreePaths} from './git-path';

/**
 * Add file contents to the index (aka staging area). If the filepath is tracked by a branch in a linked worktree, then index updates
 * will occur on the index file in the `GIT_DIR/worktrees/{branch}` directory.
 * @param filepath The relative or absolute path to add.
 * @returns {Promise<boolean>} A Promise object containing a boolean indicating whether the file was successfully staged.
 */
export const add = async (filepath: PathLike): Promise<boolean> => {
  const {dir, worktreeDir} = await getWorktreePaths(filepath.toString());
  if (!dir) return false; // not under version control

  const root = worktreeDir ? worktreeDir : dir;
  const relativePath = relative(root.toString(), filepath.toString());

  const output = await execute({
    command: 'git',
    args: ['add', relativePath],
    cwd: root.toString(),
  });
  if (output.stderr) {
    console.error(output.stderr);
    return false;
  }
  console.log(`added ${filepath.toString()} `, output.stdout);
  return true;
};
