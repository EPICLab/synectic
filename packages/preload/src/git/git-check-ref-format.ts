import {execute} from '../io';

export const checkRefFormat = async ({
  refname,
  branch = false,
}: {
  refname: string;
  branch?: boolean;
}): Promise<boolean> => {
  const output = await execute({
    command: 'git',
    args: ['check-ref-format', branch ? '--branch' : '', refname],
  });
  if (output.stderr) {
    console.error(output.stderr);
    return false;
  }
  console.log(output.stdout);
  return true;
};
