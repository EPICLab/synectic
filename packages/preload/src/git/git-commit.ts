import type {PathLike} from 'fs-extra';
import {execute} from '../io';
import {revParse} from './git-rev-parse';

/**
 * Record changes to the repository. Create a new commit containing the current contents of the
 * index and the given log message describing the changes. The new commit is a direct child of
 * HEAD, usually the tip of the current branch, and the branch is updated to point to it (unless
 * no branch is associated with the working tree, in which case HEAD is "detached" as described
 * in {@linkcode checkoutDetached}.
 * @param obj - A destructured object for named parameters.
 * @param obj.dir - The worktree root directory.
 * @param obj.message - The commit message.
 * @returns {Promise<string | undefined>} A Promise object containing the SHA-1 hash of the new
 * commit, or undefined if no commit was created.
 */
export const commit = async ({
  dir,
  message,
}: {
  dir: PathLike;
  message?: string;
}): Promise<string | undefined> => {
  const refPattern = new RegExp('\\[\\w+ (\\w+)\\]', 'g');
  const output = await execute({
    command: 'git',
    args: ['commit', message ? `-m "${message}"` : ''],
    cwd: dir.toString(),
  });

  if (output.stderr) {
    console.error(output.stderr);
    return undefined;
  }
  const sha1 = refPattern.exec(output.stdout ?? '')?.[1];
  return sha1 ? await revParse({dir: dir, args: [sha1]}) : undefined;
};
