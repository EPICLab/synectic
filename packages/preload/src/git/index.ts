export * from './git-add';
export * from './git-branch';
export * from './git-cat-file';
export * from './git-check-ref-format';
export * from './git-checkout';
export * from './git-clone';
export * from './git-commit';
export * from './git-config';
export * from './git-ignore';
export * from './git-log';
export * from './git-merge-base';
export * from './git-merge';
export * from './git-path';
export * from './git-remote';
export * from './git-reset';
export * from './git-restore';
export * from './git-rev-list';
export * from './git-rev-parse';
export * from './git-show-branch';
export * from './git-status';
export * from './git-utils';
export * from './git-version';
export * from './git-worktree';
