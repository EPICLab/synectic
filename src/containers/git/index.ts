export * from './git-branch';
export * from './git-cat-file';
export * from './git-check-ref-format';
export * from './git-checkout';
export * from './git-clone';
export * from './git-ignore';
export * from './git-log';
export * from './git-merge';
export * from './git-path';
export * from './git-rev-parse';
export * from './git-status';
export * from './git-version';
export * from './git-worktree';