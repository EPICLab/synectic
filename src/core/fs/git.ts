import * as fs from 'fs-extra';
import * as path from 'path';
// import * as util from 'util';
import * as git from 'isomorphic-git';
git.plugins.set('fs', fs);

export function getRepoRoot(p: fs.PathLike): Promise<string> {
  return git.findRoot({ filepath: p.toString() });
}

/**
 * Synchronous check for presence of .git within directory to validate Git version control.
 * @param directory A path to a file. If a URL is provided, it must use the `file:` protocol.
 * @return Boolean indicating directory under Git version control.
 */
export function isGitRepo(directory: fs.PathLike): boolean {
  const p: string = path.resolve(path.join(directory.toString(), '/.git'));
  // TODO: Handle cases where directory includes a filename which needs to be removed.
  let result: boolean;
  try {
    fs.statSync(p);
    result = true;
  } catch (error) {
    result = false;
  }
  return result;
}

/**
 * Asynchronous check for presence of .git within directory to validate Git version control.
 * @param directory A path to a file. If a URL is provided, it must use the `file:` protocol.
 * @return Boolean indicating directory under Git version control.
 */
export function isGitRepoAsync(directory: fs.PathLike): Promise<boolean> {
  const p: string = path.resolve(path.join(directory.toString(), '/.git'));
  return new Promise((resolve, _) => {
    fs.stat(p)
      .then(() => {
        resolve(true);
      })
      .catch(() => {
        resolve(false);
      });
  });
  // return fs.stat(p)
  //   .then(() => true)
  //   .catch(() => false);
}
