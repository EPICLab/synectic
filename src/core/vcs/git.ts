import * as fs from 'fs-extra';
import * as path from 'path';
// import * as util from 'util';
import * as git from 'isomorphic-git';
import * as io from '../fs/io';
import { basename } from 'path';

export * from 'isomorphic-git';

// Lim requires a function that will compare a local file against the latest
// version on a git branch, and returns boolean for diff.

export async function fetchRepo(directory: fs.PathLike) {
  await git.fetch({
    dir: directory.toString(),
    token: '02382a038d912a9fc9fd88b7feb77d5084ff30f0',
    oauth2format: 'github',
    depth: 1,
    singleBranch: false,
    tags: false
  });
  console.log('done');
}

export async function getRepoFullname(): Promise<string> {
  const value = await git.config({
    dir: '/',
    path: 'user.name'
  });
  console.log('user-name: ' + value);
  return new Promise((resolve, _) => {
    resolve('test');
  });
}

export async function checkoutFile(filepath: fs.PathLike, branch: string): Promise<string> {
  // TODO: Big optimization is possible here.
  // The 'checkout' command from isomorphic-git currently deletes and rewrites
  // everything in the project as part of a branch checkout, even when a pattern
  // is supplied. This forces us to checkout the target branch, grab the file
  // we are interested in and move it to a temporary location, switch back to
  // the original branch and return the path to the temporary file for loading
  // by the card.
  const repoRoot = await getRepoRoot(filepath);
  const currentBranch = await git.currentBranch({ dir: repoRoot, fullname: false });
  const relativePath = path.relative(repoRoot, filepath.toString());
  let targetPath = '';
  if (branch === currentBranch) {
    targetPath = path.join(repoRoot, basename(filepath.toString()));
  } else {
    targetPath = path.join(repoRoot, '/.git/tmp/', basename(filepath.toString()));
    await git.checkout({ dir: repoRoot, ref: branch, pattern: relativePath });
    await fs.move(filepath.toString(), targetPath, { overwrite: true });
    await git.checkout({ dir: repoRoot, ref: currentBranch });
  }
  return targetPath;
}

/**
 * Read remote name and URL from Git directory path.
 * Git protocol allows multiple remotes to be set per repository.
 * @param gitdir The git directory path.
 * @return A list of git.RemoteDefinition objects, each containing `remote` and `url` fields.
 */
export async function getRemotes(gitdir: fs.PathLike): Promise<git.RemoteDefinition[]> {
  return git.listRemotes({ dir: gitdir.toString() });
}

/**
 * Recursively walk up directories from path to locate Git directory path.
 * @param p A path to a file or directory that is under Git version control.
 * @param relative (Optional) Flag for generating a relative path to the Git directory path.
 * @return The absolute (or relative) path to the Git path directory.
 */
export async function getRepoRoot(p: fs.PathLike, relative?: boolean): Promise<string> {
  const rootPath = git.findRoot({ filepath: p.toString() });
  return relative ? path.relative(await rootPath, p.toString()) : rootPath;
}

/**
 * Asynchronously aggregate local and remote branches based on Git repository.
 * @param gitdir The git directory path.
 * @return A deduplicated array of all branch names for the Git repository.
 */
export async function getAllBranches(gitdir: fs.PathLike): Promise<string[]> {
  const local = await getLocalBranches(gitdir);
  const remote = await getRemoteBranches(gitdir);
  const allBranches = new Set(local.concat(remote));
  return Array.from(allBranches.values());
}

/**
 * Asynchronously aggregate local branches based on Git repository.
 * @param gitdir The git directory path.
 * @return An array of local branch names for the Git repository.
 */
export async function getLocalBranches(gitdir: fs.PathLike): Promise<string[]> {
  return git.listBranches({ dir: gitdir.toString() });
}

/**
 * Asynchronously aggregate remote branches based on Git repository.
 * @param gitdir The git directory path.
 * @return An array of remote branch names for the Git repository.
 */
export async function getRemoteBranches(gitdir: fs.PathLike): Promise<string[]> {
  return git.listBranches({ dir: gitdir.toString(), remote: 'origin' });
}

/**
 * Synchronously validate that directory is Git directory path (i.e. contains .git file).
 * @param directory A path to a file. If a URL is provided, it must use the `file:` protocol.
 * @return Boolean indicating directory is a Git directory path.
 */
export function isGitRepo(directory: fs.PathLike): boolean {
  const p: string = path.resolve(path.join(directory.toString(), '/.git'));
  // TODO: Handle cases where directory includes a filename which needs to be removed.
  let result: boolean;
  try {
    fs.statSync(p);
    result = true;
  } catch (error) {
    result = false;
  }
  return result;
}

/**
 * Asynchronously validate that directory is Git directory path (i.e. contains .git file).
 * @param directory A path to a file. If a URL is provided, it must use the `file:` protocol.
 * @return Boolean indicating directory is a Git directory path.
 */
export function isGitRepoAsync(directory: fs.PathLike): Promise<boolean> {
  const p: string = path.resolve(path.join(directory.toString(), '/.git'));
  return io.exists(p);
}
