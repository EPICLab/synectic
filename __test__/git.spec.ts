import mock from 'mock-fs';
import * as path from 'path';
import { homedir } from 'os';

import * as git from '../src/containers/git';

// const mockGitPath = path.resolve(__dirname, '__mocks__', 'gitProjMock');
// const mockGitProj = {
//   empty: {},
//   foo: {
//     bar: {
//       'untracked-file.js': 'directory is untracked by git',
//     },
//     baz: {
//       '.git': {
//         'HEAD': 'ref: refs/heads/feature/test',
//         'config': '[core]\nrepositoryformatversion = 0\nfilemode = true\nbare = false\nlogallrefupdates = true\nignorecase = true\nprecomposeunicode = true\n[remote "origin"]\nurl = git@github.com:test/test.git\nfetch = +refs / heads/*:refs/remotes/origin/*\n[branch "master"]\nremote = origin\nmerge = refs/heads/master',
//         objects: {
//           'e2': {
//             '7bb34b0807ebf1b91bb66a4c147430cde4f08f': Buffer.from([98, 108, 111, 98, 32, 50, 53, 0, 77, 121, 32, 100, 97, 116, 97, 32, 102, 105, 116, 115, 32, 111, 110, 32, 111, 110, 101, 32, 108, 105, 110, 101, 10]),
//           },
//           '42': {
//             '2a8a27eebd3798c661f2c0788dc8d6dfe597a1': `blob 26\x00My data fits on line line\n`
//           }
//         },
//         refs: {
//           heads: {
//             feature: {
//               'test': '4c40253aace4ffa46c943311d77232cb5d4ffe93'
//             },
//             'master': '4c40253aace4ffa46c943311d77232cb5d4ffe93',
//             'remote-only': 'a81c46a181052b4bbb0037b7ab192540c4234054'
//           },
//         }
//       },
//       'tracked-file.js': 'directory is tracked by git',
//       'another-file.ts': 'directory is tracked by git, but the git repo is currently in a detached HEAD state'
//     },
//     qux: {
//       '.git': {
//         'HEAD': '5862ad5c2f677a657b09fe5651693df60fb64227',
//         'config': '[core]\nrepositoryformatversion = 0\nfilemode = true\nbare = false\nlogallrefupdates = true\nignorecase = true\nprecomposeunicode = true\n[remote "origin"]\nurl = git@github.com:test/test.git\nfetch = +refs / heads/*:refs/remotes/origin/*\n[branch "master"]\nremote = origin\nmerge = refs/heads/master',
//         objects: {}
//       }
//     }
//   }
// };

// describe('git.currentBranch', () => {
//   it('currentBranch resolves to Git branch name on a tracked directory', async () => {
//     await expect(git.currentBranch({ dir: path.resolve(mockGitPath, 'foo/baz/') })).resolves.toBe('test');
//   });

//   it('currentBranch resolves to undefined on a tracked directory with detached HEAD', async () => {
//     await expect(git.currentBranch({ dir: path.resolve(mockGitPath, 'foo/qux/') })).resolves.toBeUndefined();
//   });

//   it('currentBranch fails with an error on an untracked directory', async () => {
//     await expect(git.currentBranch({ dir: path.resolve(mockGitPath, 'foo/bar/') })).rejects.toThrow(/Could not find HEAD/);
//   });
// });

// describe('git.getStatus', () => {
//   it('getStatus resolves Git status on tracked file', async () => {
//     await expect(git.getStatus(path.resolve(mockGitPath, 'foo/baz/tracked-file.js.md'))).resolves.toBe('unmodified');
//   });

//   it('getStatus resolves Git status on tracked directory', async () => {
//     await expect(git.getStatus(path.resolve(mockGitPath, 'foo/baz'))).resolves.toBe('unmodified');
//   });

//   it('getStatus resolves to undefined on untracked file', async () => {
//     mock(mockGitProj);
//     await expect(git.getStatus('foo/bar/untracked-file.js')).resolves.toBeUndefined();
//     mock.restore();
//   });

//   it('getStatus resolves to undefined on untracked directory', async () => {
//     mock(mockGitProj);
//     await expect(git.getStatus('foo/bar/')).resolves.toBeUndefined();
//     mock.restore();
//   });
// });

// describe('git.getRepoRoot', () => {
//   it('getRepoRoot resolves to Git root directory on file in tracked directory', async () => {
//     return expect(git.getRepoRoot(`${mockGitPath}/foo/baz/tracked-file.js`)).resolves.toBe(path.resolve(mockGitPath, 'foo/baz'));
//   });

//   it('getRepoRoot resolves to undefined on file in untracked directory', async () => {
//     mock(mockGitProj);
//     await expect(git.getRepoRoot(`foo/bar/untracked-file.js`)).resolves.toBeUndefined();
//     mock.restore();
//   });
// });

// describe('git.isGitRepo', () => {
//   it('isGitRepo resolves direct parent directory of .git directory to true', async () => {
//     return expect(git.isGitRepo(`${mockGitPath}/foo/baz/`)).resolves.toBe(true);
//   });

//   it('isGitRepo resolves directory path ending in .git directory to true', async () => {
//     return expect(git.isGitRepo(`${mockGitPath}/foo/baz/.git`)).resolves.toBe(true);
//   });

//   it('isGitRepo resolves file path containing an adjacent .git directory to true', async () => {
//     return expect(git.isGitRepo(`${mockGitPath}/foo/baz/another-file.ts`)).resolves.toBe(true);
//   });

//   it('isGitRepo resolves directory path without a .git directory to false', async () => {
//     return expect(git.isGitRepo(`${mockGitPath}/foo/bar/`)).resolves.toBe(false);
//   });

//   it('isGitRepo resolves nonexistent path ending in .git directory to false', async () => {
//     return expect(git.isGitRepo(`${mockGitPath}/foo/bar/.git`)).resolves.toBe(false);
//   });
// });

describe('git.extractRepoName', () => {
  it('extractRepoName resolves git://*', () => {
    expect(git.extractRepoName('git://github.com/octo-org/octo-repo')).toBe('octo-org/octo-repo');
  });

  it('extractRepoName resolves git://*.git', () => {
    expect(git.extractRepoName('git://github.com/octo-org/octo-repo.git')).toBe('octo-org/octo-repo');
  });

  it('extractRepoName resolves https://*', () => {
    expect(git.extractRepoName('https://treygriffith@bitbucket.org/bucket-org/cellar.git')).toBe('bucket-org/cellar');
  });

  it('extractRepoName resolves https://*.git', () => {
    expect(git.extractRepoName('https://gitlab.com/gitlab-org/omnibus-gitlab.git')).toBe('gitlab-org/omnibus-gitlab');
  });

  it('extractRepoName resolves ssh://*.git', () => {
    expect(git.extractRepoName('ssh://git@gitlab.com:labuser/lab-repo.git')).toBe('labuser/lab-repo');
  });

  it('extractRepoName resolves https://gist', () => {
    expect(git.extractRepoName('https://bitbucket.org/snippets/vmaric/oed9AM/hello-json-message')).toBe('vmaric/oed9AM/hello-json-message');
  });

  it('extractRepoName resolves git@gist', () => {
    expect(git.extractRepoName('git@gist.github.com:3135914.git')).toBe('3135914');
  });

  it('extractRepoName resolves git+https://*.git#tag', () => {
    expect(git.extractRepoName('git+https://github.com/octo-org/octo-repo.git#2.7.0')).toBe('octo-org/octo-repo');
  });
});

describe('git.getConfig', () => {

  beforeAll(() => {
    mock({
      [path.join(homedir(), '.gitconfig')]: mock.file({
        content: `[user]
  name = Sandy Updates
  email = supdate@oregonstate.edu
[core]
  editor = vim
  whitespace = fix,-indent-with-non-tab,trailing-space,cr-at-eol`,
      }),
      '.git/config': mock.file({
        content: `[user]
  name = Bobby Tables
  email = bdrop@oregonstate.edu
[credential]
  helper = osxkeychain
[pull]
  rebase = true
[alias]
  last = log -1 HEAD`,
      }),
    }, { createCwd: false });
  });

  afterAll(mock.restore);

  it('getConfig resolves global git-config value', async () => {

    return expect(git.getConfig('user.name', true)).resolves.toStrictEqual({ scope: 'global', value: 'Sandy Updates' });
  });

  it('getConfig resolves local git-config value', async () => {
    return expect(git.getConfig('user.name')).resolves.toStrictEqual({ scope: 'local', value: 'Bobby Tables' });
  });

  it('getConfig checks both scopes for git-config value', async () => {
    return expect(git.getConfig('core.editor')).resolves.toStrictEqual({ scope: 'global', value: 'vim' });
  });

  it('getConfig resolves none when no git-config value is found', async () => {
    return expect(git.getConfig('commenter.name')).resolves.toStrictEqual({ scope: 'none' });
  });
});

describe('git.setConfig', () => {

  beforeEach(() => {
    mock({
      [path.join(homedir(), '.gitconfig')]: mock.file({
        content: `[user]
  name = Sandy Updates
  email = supdate@oregonstate.edu
[core]
  editor = vim`,
      }),
      '.git/config': mock.file({
        content: `[user]
  name = Bobby Tables
  email = bdrop@oregonstate.edu
[credential]
  helper = osxkeychain`,
      }),
    }, { createCwd: false });
  });

  afterAll(mock.restore);

  it('setConfig adds value to global git-config', async () => {
    const updated = await git.setConfig('global', 'user.username', 'supdate');
    mock.restore(); // required to prevent snapshot rewriting because of file watcher race conditions in Jest
    expect(updated).toMatchSnapshot();
  });

  it('setConfig updates values to local git-config', async () => {
    const updated = await git.setConfig('local', 'user.name', 'John Smith');
    mock.restore(); // required to prevent snapshot rewriting because of file watcher race conditions in Jest
    expect(updated).toMatchSnapshot();
  });

  it('setConfig deletes values from local git-config', async () => {
    const updated = await git.setConfig('local', 'user.name', undefined);
    mock.restore(); // required to prevent snapshot rewriting because of file watcher race conditions in Jest
    expect(updated).toMatchSnapshot();
  });
});