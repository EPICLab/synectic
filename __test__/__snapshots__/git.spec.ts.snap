// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git.extractRepo extractRepo resolves a new Git repository to { new repo, AddRepoAction action } 1`] = `
Object {
  "corsProxy": "https://cors-anywhere.herokuapp.com/",
  "id": undefined,
  "name": "test/test",
  "oauth": "github",
  "password": "",
  "refs": Array [
    "feature/test",
  ],
  "token": "",
  "url": Object {
    "hash": "",
    "href": "git@github.com:test/test.git",
    "pathname": "/test/test.git",
    "port": null,
    "protocol": "ssh",
    "protocols": Array [],
    "query": Object {},
    "resource": "github.com",
    "search": "",
    "user": "git",
  },
  "username": "",
}
`;

exports[`git.extractRepo extractRepo resolves a new Git repository to { new repo, AddRepoAction action } 2`] = `
Object {
  "id": undefined,
  "repo": Object {
    "corsProxy": "https://cors-anywhere.herokuapp.com/",
    "id": undefined,
    "name": "test/test",
    "oauth": "github",
    "password": "",
    "refs": Array [
      "feature/test",
    ],
    "token": "",
    "url": Object {
      "hash": "",
      "href": "git@github.com:test/test.git",
      "pathname": "/test/test.git",
      "port": null,
      "protocol": "ssh",
      "protocols": Array [],
      "query": Object {},
      "resource": "github.com",
      "search": "",
      "user": "git",
    },
    "username": "",
  },
  "type": 13,
}
`;
